text/0009-client-api.md

```diff
@@ -0,0 +1,154 @@
+    * `block_finalized`
+    * `block_added`
+    * `deploy_added`
+    * `deploy_requeued`
+    * `deploy_discarded`
+    * `deploy_finalized`
+    * `deploy_processed`
+    * `deploy_orphaned`
+* alternatively, `subscription=all` implies subscribing to all available event types
+* `accounts=[comma-separated list of hex-encoded accounts]` to filter by the listed accounts
+* `deploy_hashes=[comma-separated list of hex-encoded deploy hashes]` to filter by the listed deploys
+
+The events corresponding to these subscriptions are as close as possible to those currently defined
+in [info.proto](https://github.com/CasperLabs/CasperLabs/blob/cd2e80286d4172c9d2e73a2c1f1271b2961a7527/protobuf/io/casperlabs/casper/consensus/info.proto#L91-L161).
+
+The first message returned will be the API version.  All subsequent ones will be JSON-encoded events.
+
+### Use case
+
+It's anticipated that the Rust client would be able to send a new deploy using the /deploy "put" RPC, while also having
+subscribed to events filtered on that deploy's ID.
+
+In this way, the client's `put-deploy` and `send-deploy` subcommands could accept arguments specifying that the client
+wait until the deploy has transitioned to certain state.  Internally, the client will consume events (possibly reporting
+each) until the requested deploy state has been reached, at which point the client can disconnect.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+* Clients will not be able to receive as rich information in response to requests for information as they currently do.
+
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* Plain JSON requests and responses via a REST API is a viable alternative to JSON-RPCs.  There
+appears to be little difference in usability or implementation effort, although other projects in
+the blockchain space use JSON-RPC, so perhaps there would be some slight benefit for us to follow
+suit.
+
+* Websockets could be used instead of SSEs.  These should work for us, but they provide full-duplex
+connections, allowing back and forth messaging between the client and server.  Our use case is
+simpler: the client specifies what it wishes to subscribe to in the GET request, and all further
+traffic is from the node to the client.  As such, SSEs seems to be the simpler, better choice.
+
+* Multiple event streams could be provided by grouping events into categories.  It doesn't seem like
+this would be of any benefit, and if it results in a large increase of HTTP connections, could
+instead be detrimental.
+
+* Node could persist all ancillary information similarly to what is done in the Scala codebase.
+This would increase the complexity of node and decrease the performance.  It would also consume
+memory and disk resources.  It seems like a better solution to offload that to a separate process
+designed specifically to handle that information.
```

Comment:
My strong preference is to offload as much data as possible to some manner of middleware event stream consumer / data aggregator.

Having said that, we may need to store some kind of executed deploy summary so that we can answer basic questions about deploy disposition, as a short term measure until such a time as we can offer a middleware implementation.

Reply:
The CEP has been updated to reflect this now.